Bucketeer/Sources/Internal/Database/SQLite/SQLite.swift

```diff
@@ -5,31 +5,37 @@ final class SQLite {
     let path: String
     let pointer: OpaquePointer
     let logger: Logger?
+    // access SQLite on one serial queue to prevent `database is locked` or database corrupt
+    private static let dbQueue = DispatchQueue(label: "io.bucketeer.SQLite")
 
     init(path: String, logger: Logger?) throws {
         self.path = path
         self.logger = logger
-
-        sqlite3_shutdown()
-        sqlite3_initialize()
-
-        var _pointer: OpaquePointer?
-        let flags = SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX
-        let result = sqlite3_open_v2(path, &_pointer, flags, nil)
-        if result != SQLITE_OK {
-            guard let pointer = _pointer else {
+        pointer = try Self.dbQueue.sync {
+            var _pointer: OpaquePointer?
+            let flags = SQLITE_OPEN_CREATE | SQLITE_OPEN_READWRITE | SQLITE_OPEN_FULLMUTEX
+            let result = sqlite3_open_v2(path, &_pointer, flags, nil)
+            if result != SQLITE_OK {
+                guard let dbConnection = _pointer else {
+                    throw Error.unknown
+                }
+                throw Error.failedToOpen(.init(errorMessage: dbConnection.readErrorMessage(), result: result))
+            }
+            guard let dbConnection = _pointer else {
                 throw Error.unknown
             }
-            throw Error.failedToOpen(.init(pointer: pointer, result: result))
-        }
-        guard let pointer = _pointer else {
-            throw Error.unknown
+            return dbConnection
         }
-        self.pointer = pointer
+        // allow concurrent reads and writes https://www.sqlite.org/walformat.html
+        // its nice to have
+        try exec(query: "PRAGMA journal_mode = WAL")
     }
-
+    
     deinit {
-        sqlite3_close(pointer)
+        let dbConnection = pointer
+        Self.dbQueue.async {
+            sqlite3_close_v2(dbConnection)
```

Comment:
You commented that we need a connection pool because closing and opening the db connection quickly could get errors or corrupt the data. Will this fix the issue by closing it asynchronously?

Also, I noticed when opening the connection is using `sqlite3_open_v2`, and in the deinit, you are now using the `sqlite3_close_v2`.
Is this another issue using different versions?

Reply:
Yes, after moving all database operations to the serial queue, no more data corruption.
So we don't need to create a connection pool for now.

Reply:
I didn't see a problem with `sqlite3_close`, but because we are using `sqlite3_open_v2`. 
It is better to use the same close-function version